community/kernel/src/main/java/org/neo4j/kernel/impl/api/state/NodeStateImpl.java

```diff
@@ -41,8 +40,141 @@
 import static java.util.Collections.emptyIterator;
 import static org.neo4j.collection.primitive.Primitive.intSet;
 
-public class NodeStateImpl extends PropertyContainerStateImpl implements NodeState
+class NodeStateImpl extends PropertyContainerStateImpl implements NodeState
 {
+    static final NodeState EMPTY = new NodeState()
+    {
+        @Override
+        public Iterator<StorageProperty> addedProperties()
+        {
+            return emptyIterator();
+        }
+
+        @Override
+        public Iterator<StorageProperty> changedProperties()
+        {
+            return emptyIterator();
+        }
+
+        @Override
+        public Iterator<Integer> removedProperties()
+        {
+            return emptyIterator();
+        }
+
+        @Override
+        public Iterator<StorageProperty> addedAndChangedProperties()
+        {
+            return emptyIterator();
+        }
+
+        @Override
+        public Iterator<StorageProperty> augmentProperties( Iterator<StorageProperty> iterator )
+        {
+            return iterator;
+        }
+
+        @Override
+        public void accept( PropertyContainerState.Visitor visitor )
+        {
+        }
+
+        @Override
+        public ReadableDiffSets<Integer> labelDiffSets()
+        {
+            return ReadableDiffSets.Empty.instance();
+        }
+
+        @Override
+        public int augmentDegree( Direction direction, int degree )
+        {
+            return degree;
+        }
+
+        @Override
+        public int augmentDegree( Direction direction, int degree, int typeId )
+        {
+            return degree;
+        }
+
+        @Override
+        public void accept( NodeState.Visitor visitor )
+        {
+        }
+
+        @Override
+        public PrimitiveIntSet relationshipTypes()
+        {
+            return intSet();
+        }
+
+        @Override
+        public long getId()
+        {
+            throw new UnsupportedOperationException( "id not defined" );
+        }
+
+        @Override
+        public boolean hasPropertyChanges()
+        {
+            return false;
+        }
+
+        @Override
+        public boolean hasRelationshipChanges()
+        {
+            return false;
+        }
+
+        @Override
+        public StorageProperty getChangedProperty( int propertyKeyId )
+        {
+            return null;
+        }
+
+        @Override
+        public StorageProperty getAddedProperty( int propertyKeyId )
+        {
+            return null;
+        }
+
+        @Override
+        public boolean isPropertyChangedOrRemoved( int propertyKey )
+        {
+            return false;
+        }
+
+        @Override
+        public boolean isPropertyRemoved( int propertyKeyId )
+        {
+            return false;
+        }
+
+        @Override
+        public PrimitiveLongIterator getAddedRelationships( Direction direction )
+        {
+            return null;
```

Comment:
next 4 methods should return an empty iterator instead, to make it consistent across methods that return iterators at least

Reply:
I didn't want to change/break existing contracts, but I agree it should be fixed.